propolis/src/block/mod.rs

```diff
@@ -188,32 +298,211 @@ impl Notifier {
         let mut inner = self.notifier.lock().unwrap();
         *inner = val;
     }
+
+    /// Stop accepting requests from the block device.
+    ///
+    /// Given there might be in-flight requests being handled by the backend,
+    /// the `Notifier::paused` method returns a future indicating
+    /// when the pause operation is complete.
+    pub fn pause(&self) {
+        // Stop responding to any requests
+        assert!(!self.paused.swap(true, Ordering::Release));
```

Comment:
style nit: Could the result from the swap be stored in a local prior to the `assert!()`?  I think it makes it easier to read though the logic while consider the asserts just additional safety around the expected invariants.